daemon/containerd/mount.go

```diff
@@ -0,0 +1,51 @@
+package containerd
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/containerd/containerd/mount"
+	"github.com/docker/docker/container"
+	"github.com/sirupsen/logrus"
+)
+
+// Mount mounts and sets container base filesystem
+func (i *ImageService) Mount(ctx context.Context, container *container.Container) error {
+	snapshotter := i.client.SnapshotService(i.snapshotter)
+	mounts, err := snapshotter.Mounts(ctx, container.ID)
+	if err != nil {
+		return err
+	}
+
+	tempDir := os.TempDir()
+
+	root, err := os.MkdirTemp(tempDir, fmt.Sprintf("%s_rootfs-mount", container.ID))
```

Comment:
Looking at this a bit; so, this is the mount point on the host for the container's rootfs? Maybe I'm very much off the mark (WELP!), but wondering;

- It's currently using the host's `TempDir()` - is that the expected location for this? (many tools - and users - may be interacting with the tempdir. So a user may be (idk) `rm -rf /tmp` or other things
- `TempDir()` may be on a different drive as the snapshotter (?)
- Do we need specific SELinux and/or AppArmor rules for these paths?
- Does the path have to be in some location owned / managed by either `dockerd` or `containerd` ?
- What path(s) does the non-snapshotter code use for these?

Reply:
The `Mount` operation (in the graph driver implementation) has two purposes:

* to temporarily mount the container FS and do random operations on it:
  * create the working directory on build (classic): https://github.com/moby/moby/blob/db11ed27bde60cbedcdbcc46b01936adec6d3e22/daemon/workdir.go#L14 
  * `stat`ing a path inside the container's FS: https://github.com/moby/moby/blob/db11ed27bde60cbedcdbcc46b01936adec6d3e22/daemon/archive_windows.go#L30 and https://github.com/moby/moby/blob/db11ed27bde60cbedcdbcc46b01936adec6d3e22/daemon/create_unix.go#L24 and https://github.com/moby/moby/blob/db11ed27bde60cbedcdbcc46b01936adec6d3e22/daemon/oci_windows.go#L86

`All` of these operations are temporary (they all are followed by a `defer ..Unmount`.

Some of these operations still make sense with the containerd image store, hence the `TempDir` here.

_Note_: @corhere made a PR recently adding a nicer way to temporarily mount a container filesystem (https://github.com/moby/moby/pull/44210), we might want to take a look if it's possible to replace all these `Mount` calls with his implementation.

The other use case for `Mount` is effectively mounting the container rootfs when preparing it:
* https://github.com/moby/moby/blob/db11ed27bde60cbedcdbcc46b01936adec6d3e22/daemon/daemon_windows.go#L499
* https://github.com/moby/moby/blob/db11ed27bde60cbedcdbcc46b01936adec6d3e22/daemon/daemon_unix.go#L1397

These operations don't make sense with containerd, we let containerd create the rootfs for the container. These codepaths are no longer taken when the containerd image store feature is enabled. 

Last note: now that the image service holds the `Mount` and `Unmount` operations we could remove them from the daemon itself, we can do this in a followup

Reply:
Thanks for explaining! So, perhaps;

1. We should add a comment around this code to outline that these mounts are temporary (and why), as it may not be immediately obvious
2. Still thinking if the `os.TempDir()` is the right location (even if temporary).
    - Wondering what a suitable alternative would be; _possibly_ we could something inside `/var/lib/docker/tmp` for this, or another directory underneath `/var/lib/docker`
    - IIUC from a conversation we had, it's non-trivial to know what the containerd's snapshotter storage location will be (so, we don't have any guarantees that the mount will be on the same filesystem) (probably "ok" for most cases, but wondering if there _may_ be cases where that's an issue).

Reply:
> 2. Still thinking if the os.TempDir() is the right location (even if temporary).
    Wondering what a suitable alternative would be; possibly we could something inside /var/lib/docker/tmp for this, or another directory underneath /var/lib/docker

IIRC `os.TempDir()` is already under `/var/lib/docker/...` because we set the `TMPDIR`: 
https://github.com/moby/moby/blob/a0e8922d13de8ad5570f54cd863dd5c9bfade581/daemon/daemon.go#L777

and TempDir seems to pick it up:
> On Unix systems, it returns $TMPDIR if non-empty, else /tmp.
https://pkg.go.dev/os#TempDir

Reply:
> IIRC os.TempDir() is already under /var/lib/docker/... because we set the TMPDIR:

Ah, good call. Hmm.. also something we should probably make more transparent (at least I got confused; very likely others would as well)

Reply:
Sure, but that's out of scope for this PR? Or do you want us to fix all of this now?

Reply:
Added some comments, please review